rewrite-core/src/main/java/org/openrewrite/GitRemote.java

```diff
@@ -0,0 +1,178 @@
+
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Value
+public class GitRemote {
+    Service service;
+    String url;
+    String origin;
+    String path;
+
+    @Nullable
+    String organization;
+
+    String repositoryName;
+
+    public enum Service {
+        GitHub,
+        GitLab,
+        Bitbucket,
+        BitbucketCloud,
+        AzureDevOps,
+        Unknown
+    }
+
+    public static class Parser {
+        private final Map<String, Service> origins;
+
+        public Parser() {
+            origins = new LinkedHashMap<>();
+            origins.put("github.com", Service.GitHub);
+            origins.put("gitlab.com", Service.GitLab);
+            origins.put("bitbucket.org", Service.BitbucketCloud);
+            origins.put("dev.azure.com", Service.AzureDevOps);
+            origins.put("ssh.dev.azure.com", Service.AzureDevOps);
+        }
+
+        public Parser registerRemote(Service service, String origin) {
+            if (origin.startsWith("https://") || origin.startsWith("http://") || origin.startsWith("ssh://")) {
+                origin = new Parser.HostAndPath(origin).concat();
+            }
+            if (origin.contains("@")) {
+                origin = new Parser.HostAndPath("https://" + origin).concat();
+            }
+            if (service == Service.Unknown) {
+                // Do not override a known with an unknown service
+                origins.putIfAbsent(origin, service);
+            } else {
+                origins.put(origin, service);
+            }
+            return this;
+        }
+
+        public GitRemote parse(String url) {
+            Parser.HostAndPath hostAndPath = new Parser.HostAndPath(url);
+
+            String origin = hostAndPath.host;
+            Service service = origins.get(origin);
+            if (service == null) {
+                for (String maybeOrigin : origins.keySet()) {
+                    if (hostAndPath.concat().startsWith(maybeOrigin)) {
+                        service = origins.get(maybeOrigin);
+                        origin = maybeOrigin;
+                        break;
+                    }
+                }
+            }
+
+            if (service == null) {
+                // If we cannot find a service, we assume the last 2 path segments are the organization and repository name
+                service = Service.Unknown;
+                String hostPath = hostAndPath.concat();
+                String[] segments = hostPath.split("/");
+                if (segments.length <= 2) {
+                    origin = null;
+                } else {
+                    origin = Arrays.stream(segments, 0, segments.length - 2).collect(Collectors.joining("/"));
+                }
+            }
+
+            String repositoryPath = hostAndPath.repositoryPath(origin);
+
+            switch (service) {
+                case AzureDevOps:
```

Comment:
missing a break statement. is that intentional?

Reply:
no, I missed it. I initially returned in the switch..